gossip/bloom.go

```diff
@@ -0,0 +1,108 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package gossip
+
+import (
+	"encoding/binary"
+	"hash"
+	"sync"
+
+	bloomfilter "github.com/holiman/bloomfilter/v2"
+)
+
+func NewBloomFilterFromBytes(bytes []byte) (*BloomFilter, error) {
+	b, err := NewBloomFilter()
+	if err != nil {
+		return nil, err
+	}
+
+	return b, b.bloomFilter.UnmarshalBinary(bytes)
+}
+
+func NewBloomFilter() (*BloomFilter, error) {
+	return newBloomFilter(1024, 4)
+}
+
+func newBloomFilter(m, k uint64) (*BloomFilter, error) {
+	bloomFilter, err := bloomfilter.New(m, k)
+	if err != nil {
+		return nil, err
+	}
+
+	return &BloomFilter{
+		bloomFilter: bloomFilter,
+		m:           m,
+		k:           k,
+	}, nil
+}
+
+// ResetBloomFilterIfNeeded resets a bloom filter if it has a ratio of filled elements beyond [refreshRatio].
+// Returns true if the bloom filter was reset.
+func ResetBloomFilterIfNeeded(bloomFilter *BloomFilter, refreshRatio float64) bool {
+	bloomFilter.lock.Lock()
+	defer bloomFilter.lock.Unlock()
+
+	filled := bloomFilter.bloomFilter.PreciseFilledRatio()
+
+	// only refresh the bloom filter if it breaches the refresh ratio
+	if filled < refreshRatio {
+		return false
+	}
+
+	// it's not possible for this to error assuming that the original
+	// bloom filter's parameters were valid
+	fresh, _ := bloomfilter.New(bloomFilter.m, bloomFilter.k)
+
+	bloomFilter.bloomFilter = fresh
+	return true
+}
+
+// BloomFilter is a bloom filter that is periodically refreshed when it breaches a given
+// threshold of items on Adds
+type BloomFilter struct {
```

Comment:
is this comment outdated since we now supply a function to reset it as opposed to doing so automatically?

Reply:
nit: should we move the struct definition to the top of the file above the constructors?

Reply:
do we need this wrapper of the underlying bloom filter type? currently we just add the `m` and `k` parameters, which are hardcoded internally to this package and a lock, which I'm not sure if we need on this type itself or if that could be handled by the caller.

Reply:
Removed this wrapper